include/deal.II/matrix_free/operators.h

```diff
@@ -332,6 +333,13 @@ namespace MatrixFreeOperators
     get_matrix_diagonal_inverse() const;
 
     /**
+     * Get read access to the diagonal of this operator.
+     */
+    const std::shared_ptr<DiagonalMatrix<VectorType> > &
+    get_matrix_diagonal() const;
```

Comment:
Is there a reason you return the shared_ptr instead of a const ref? That would make it less likely that people store a reference for longer.

Reply:
We use the shared pointer in other contexts, e.g. Chebyshev preconditioners and smoothers. A shared pointer in such a context is actually preferable over the reference in my opinion: The shared pointer makes sure ownership is handed over to the preconditioner in case the matrix builds a new diagonal, whereas a reference (that is stored as a reference or pointer) would be dangling. In any case, we never modify the entries in the diagonal without creating a new one, so I see it as the user's responsibility to act wisely with the shared pointer. (For example, I could imagine to set one of the diagonals to zero that come out of the `LaplaceMatrix` when I don't need that and want to save memory - but that is then solely my responsibility as a user.)

Reply:
to add to what Martin says, we switched to shared pointers some time ago to also make Chebyshev smoother more flexible in switching between different preconditioners (Jacobi is only one of them).

Reply:
Wouldn't a unique pointer be more appropriate to signal that the caller it taking over ownership and that we are no longer modifying the contents?

Reply:
> Wouldn't a unique pointer be more appropriate

i think it's more limiting. I can imagine cases where one uses diagonal in a preconditioner but also wants to do something it with elsewhere in the code. I believe shared pointer is the right choice here.